battleground_viewer/src/construct_render/construct_render.rs

```diff
@@ -10,198 +12,165 @@ use battleground_construct::Construct;
 }
 
 impl ConstructRender {
-    pub fn new(context: &Context) -> Self {
-        let mut static_geometries = vec![];
+    pub fn new() -> Self {
+        let static_meshes = MeshGeometry::new(|pass| match pass {
+            RenderPass::BaseScene | RenderPass::NonEmmisivesDepth => true,
+            _ => false,
+        });
+        let base_primitives = PrimitiveGeometry::new(|pass| match pass {
+            RenderPass::ShadowMaps | RenderPass::BaseScene | RenderPass::NonEmmisivesDepth => true,
+            _ => false,
+        });
+        let emissive_primitives = PrimitiveGeometry::new(|pass| match pass {
+            RenderPass::ShadowMaps | RenderPass::BaseScene => true,
+            _ => false,
+        });
+        let glow_primitives = PrimitiveGeometry::new(|pass| match pass {
+            RenderPass::Emmisives => true,
+            _ => false,
+        });
+        let fence_primitives = PrimitiveGeometry::new(|pass| match pass {
+            RenderPass::Fences => true,
+            _ => false,
+        });
+        let overlay_primitives = PrimitiveGeometry::new(|pass| match pass {
+            RenderPass::BaseScene => true,
+            _ => false,
+        });
+
+        ConstructRender {
+            static_meshes,
+            base_primitives,
+            emissive_primitives,
+            glow_primitives,
+            fence_primitives,
+            overlay_primitives,
+            effects: Default::default(),
+        }
+    }
+
+    fn renderables(&self) -> Vec<&dyn RenderableGeometry> {
+        let mut result = vec![
+            &self.static_meshes as &dyn RenderableGeometry,
+            &self.base_primitives as &dyn RenderableGeometry,
+            &self.emissive_primitives as &dyn RenderableGeometry,
+            &self.glow_primitives as &dyn RenderableGeometry,
+            &self.fence_primitives as &dyn RenderableGeometry,
+            &self.overlay_primitives as &dyn RenderableGeometry,
+        ];
+        for effect in self.effects.values() {
+            result.push(effect.as_renderable_geometry());
+        }
+        result
+    }
+
+    fn renderables_mut(&mut self) -> Vec<&mut dyn RenderableGeometry> {
+        let mut result = vec![
+            &mut self.static_meshes as &mut dyn RenderableGeometry,
+            &mut self.base_primitives as &mut dyn RenderableGeometry,
+            &mut self.emissive_primitives as &mut dyn RenderableGeometry,
+            &mut self.glow_primitives as &mut dyn RenderableGeometry,
+            &mut self.fence_primitives as &mut dyn RenderableGeometry,
+            &mut self.overlay_primitives as &mut dyn RenderableGeometry,
+        ];
+        for effect in self.effects.values_mut() {
+            result.push(effect.as_renderable_geometry_mut());
+        }
+        result
+    }
+
+    fn do_for_renderables<'a>(&'a self, mut f: impl FnMut(&'a dyn RenderableGeometry) -> ()) {
+        for r in self.renderables() {
```

Comment:
So... Here we get the drawbacks of internal iteration and the drawbacks of external iteration. Advice on which form is more useful here (from both a ergonomic/convention pov and performance pov) would be appreciated.

Reply:
It was an interesting (and very educational) exercise!

Let's drop `do_for_renderables`. I'm not exactly clear on how iterator chaining is any better than the for loop, but I'm not at all well-versed in Rust performance tradeoffs 😁